src/oidc_child/oidc_child.c

```diff
@@ -0,0 +1,1396 @@
+                          "Failed to get user identifier string.\n");
+                }
+            } else {
+                DEBUG(SSSDBG_CRIT_FAILURE,
+                      "Failed to read attribute [id] from userinfo data.\n");
+            }
+        }
+    }
+
+    if (user_identifier == NULL) {
+        DEBUG(SSSDBG_CRIT_FAILURE,
+              "No attribute to identify the user found.\n");
+    } else {
+        DEBUG(SSSDBG_CONF_SETTINGS, "User sub: [%s].\n", user_identifier);
+    }
+
+    return user_identifier;
+}
+
+static struct devicecode_ctx *get_dc_ctx(TALLOC_CTX *mem_ctx,
+                                         bool libcurl_debug, const char *ca_db,
+                                         const char *issuer_url,
+                                         const char *issuer,
+                                         const char *device_auth_endpoint,
+                                         const char *token_endpoint,
+                                         const char *userinfo_endpoint,
+                                         const char *jwks_uri, const char *scope)
+{
+    struct devicecode_ctx *dc_ctx = NULL;
+    int ret;
+
+    dc_ctx = talloc_zero(mem_ctx, struct devicecode_ctx);
+    if (dc_ctx == NULL) {
+        DEBUG(SSSDBG_OP_FAILURE, "Failed to allocate memory for results.\n");
+        ret = ENOMEM;
+        goto done;
+    }
+
+    dc_ctx->libcurl_debug = libcurl_debug;
+
+    if (ca_db != NULL) {
+        dc_ctx->ca_db = talloc_strdup(dc_ctx, ca_db);
+        if (dc_ctx->ca_db == NULL) {
+            DEBUG(SSSDBG_OP_FAILURE, "Failed to copy CA DB path.\n");
+            ret = ENOMEM;
+            goto done;
+        }
+    }
+
+    if (issuer_url != NULL) {
+        ret = get_openid_configuration(dc_ctx, issuer_url);
+        if (ret != EOK) {
+            DEBUG(SSSDBG_OP_FAILURE, "Failed to get openid configuration.\n");
+            goto done;
+        }
+
+        ret = parse_openid_configuration(dc_ctx);
+        if (ret != EOK) {
+            DEBUG(SSSDBG_OP_FAILURE, "Failed to parse openid configuration.\n");
+            goto done;
+        }
+    } else if (issuer != NULL) {
+        ret = get_issuer(dc_ctx, issuer);
+        if (ret != EOK) {
+            DEBUG(SSSDBG_OP_FAILURE, "Failed to get issuer configuration.\n");
+            goto done;
+        }
+    } else if (device_auth_endpoint != NULL && token_endpoint != NULL) {
+        ret = set_endpoints(dc_ctx, device_auth_endpoint, token_endpoint,
+                            userinfo_endpoint, jwks_uri, scope);
+        if (ret != EOK) {
+            DEBUG(SSSDBG_OP_FAILURE, "Failed to set endpoints.\n");
+            goto done;
+        }
+    } else {
+        DEBUG(SSSDBG_CRIT_FAILURE, "Missing issuer information.\n");
+        ret = EINVAL;
+        goto done;
+    }
+
+done:
+    if (ret != EOK) {
+        talloc_free(dc_ctx);
+        dc_ctx = NULL;
+    }
+    return dc_ctx;
+}
+
+int main(int argc, const char *argv[])
```

Comment:
This is quite large, can you split it to multiple smaller functions?

Reply:
Please have a look at the updated version.

Reply:
Much better, thanks.